torch/csrc/jit/passes/concat_opt.cpp

```diff
@@ -0,0 +1,487 @@
+    cat_out_empty->insertBefore(node);
+
+    // For every input to this `cat` node:
+    //   * Create a slice of `cat` output buffer.
+    auto cat_out_value = cat_out_empty->output();
+    auto cat_inp_list = node->input(0)->node();
+    int start_idx = 0;
+    auto start = graph_->insertConstant(start_idx);
+    for (auto cat_inp : cat_inp_list->inputs()) {
+      // Create a slice of the cat output buffer that correspond to
+      // this input size and position in the output.
+      auto cat_inp_tensor_type =
+          dynamic_cast<TensorType*>(cat_inp->type().get());
+      TORCH_INTERNAL_ASSERT(cat_inp_tensor_type);
+      TORCH_INTERNAL_ASSERT(cat_inp_tensor_type->dim());
+      auto cat_inp_tensortype_sizes = cat_inp_tensor_type->sizes();
+      int end_idx = start_idx + *cat_inp_tensortype_sizes[cat_dim_value];
+      auto end = graph_->insertConstant(end_idx);
+
+      auto slice = graph_->create(
+          slice_symbol, {cat_out_value, cat_dim, start, end, one});
+      GRAPH_UPDATE("Inserting\n", *slice, "before\n", *node);
+      slice->insertBefore(node);
+      slices_added_.push_back(slice);
+
+      // Insert a copy from this input to the output slice.
+      auto copy = graph_->create(copy_symbol, {slice->output(), cat_inp});
+      GRAPH_UPDATE("Inserting\n", *copy, "before\n", *node);
+      copy->insertBefore(node);
+      copies_added_.push_back(copy);
+
+      start_idx = end_idx;
+      start = end;
+    }
+
+    // Replace the uses of `cat` node with the cat output buffer.
+    replace_uses_with_[node->output()] = cat_out_value;
+    nodes_to_remove_.insert(node);
+  }
+
+  bool shapeIsKnown(Value* v) {
+    if (v->type()->cast<TensorType>()) {
+      if (!v->isCompleteTensor()) {
+        return false;
+      }
+      if (*v->type()->castRaw<TensorType>()->dim() == 0) {
+        return false;
+      }
+    }
+    return true;
+  }
+  bool allShapesAreKnown(Node* node) {
+    // TODO: Relax the checks to support dynamic shapes
+    for (Value* input : node->inputs()) {
+      if (!shapeIsKnown(input)) {
+        return false;
+      }
+    }
+    for (Value* output : node->outputs()) {
+      if (!shapeIsKnown(output)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  void cleanupExpandedCatOps() {
+    for (auto it : replace_uses_with_) {
+      GRAPH_UPDATE(
+          "Replacing uses of\n",
+          *it.first->node(),
+          "with\n",
+          *it.second->node());
+      it.first->replaceAllUsesWith(it.second);
+    }
+    for (auto n : nodes_to_remove_) {
+      removeCatNodeFromGraph(n);
+    }
+  }
+
+  void moveBefore(Node* node, Node* before) {
+    // In order to move a node before another node, we need to move
+    // all the nodes it depends on as well.
+    for (auto inp : node->inputs()) {
+      moveBefore(inp->node(), before);
+    }
+    node->moveBefore(before);
+  }
+
+  // Reuse buffers in copies wherever possible.
```

Comment:
Hmm, I'm having a little bit of a difficult time understanding this one. copy_ has broadcasting behavior, we might have to be a little careful. Do you mind rewriting the example you have here as python code / explaining a bit more ?

Reply:
Added some more explanation for this. PTAL.